File: modules/authentication/pkg/javascript/examples/basic.js
/**
 * Basic Authentication Example
 *
 * This example demonstrates the complete authentication flow:
 * - Network authentication
 * - Discovery service authentication
 * - Domain access
 * - State persistence
 */

import { Client } from '../dist/index.js';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config({ path: '../../.env' });

async function main() {
  console.log('=== Auki Authentication Example ===\n');

  // Configuration
  const config = {
    apiUrl: process.env.API_URL || 'https://api.aukiverse.com',
    refreshUrl: process.env.REFRESH_URL || 'https://api.aukiverse.com/user/refresh',
    ddsUrl: process.env.DDS_URL || 'https://dds.posemesh.org',
    clientId: process.env.CLIENT_ID || 'javascript-sdk',
    refreshThresholdMs: parseInt(process.env.REFRESH_THRESHOLD_MS || '300000')
  };

  console.log('Configuration:');
  console.log(`  API URL: ${config.apiUrl}`);
  console.log(`  Refresh URL: ${config.refreshUrl}`);
  console.log(`  DDS URL: ${config.ddsUrl}`);
  console.log(`  Client ID: ${config.clientId}`);
  console.log(`  Refresh threshold: ${config.refreshThresholdMs}ms\n`);

  // Credentials
  const credentials = {
    type: 'email',
    email: process.env.EMAIL,
    password: process.env.PASSWORD
  };

  if (!credentials.email || !credentials.password) {
    console.error('❌ EMAIL and PASSWORD must be set in .env file');
    process.exit(1);
  }

  console.log('Using Email/Password authentication');
  console.log(`  Email: ${credentials.email}\n`);

  // Create the authentication client
  const client = await Client.create(config);
  client.setCredentials(credentials);

  // Check initial authentication state
  console.log('Initial authentication state:', client.isAuthenticated() ? 'Authenticated' : 'Not authenticated');
  console.log();

  try {
    // Step 1: Authenticate to the Auki network
    console.log('=== Step 1: Network Authentication ===');
    const networkToken = await client.authenticate();
    console.log('✓ Network authentication successful!');
    console.log(`  Token expires at: ${networkToken.expires_at}`);
    console.log();

    // Step 2: Authenticate to Discovery service
    console.log('=== Step 2: Discovery Authentication ===');
    const discoveryToken = await client.authenticateDiscovery();
    console.log('✓ Discovery authentication successful!');
    console.log(`  Token expires at: ${discoveryToken.expires_at}`);
    console.log();

    // Step 3: Get domain access
    const domainId = process.env.DOMAIN_ID;
    if (!domainId) {
      console.log('⚠ DOMAIN_ID not set in .env, skipping domain access step');
    } else {
      console.log('=== Step 3: Domain Access ===');
      console.log(`Requesting access to domain: ${domainId}`);

      const domainAccess = await client.getDomainAccess(domainId);
      console.log('✓ Domain access granted!');
      console.log(`  Domain ID: ${domainAccess.id}`);
      console.log(`  Domain Name: ${domainAccess.name}`);
      console.log(`  Server URL: ${domainAccess.domain_server.url}`);
      console.log(`  Server Name: ${domainAccess.domain_server.name}`);
      console.log(`  Server Region: ${domainAccess.domain_server.cloud_region}`);
      console.log(`  Location: ${domainAccess.domain_server.latitude}, ${domainAccess.domain_server.longitude}`);
      console.log(`  Access token: ${domainAccess.access_token.substring(0, 50)}...`);
      console.log(`  Token expires at: ${domainAccess.expires_at}`);
      console.log();
    }

    // Display final state
    console.log('=== Final State ===');
    console.log(`Is authenticated: ${client.isAuthenticated()}`);

    const networkTokenCached = client.getNetworkToken();
    if (networkTokenCached) {
      console.log('\nNetwork Token:');
      console.log(`  Token: ${networkTokenCached.token.substring(0, 50)}...`);
      console.log(`  Expires at: ${networkTokenCached.expires_at}`);
    }

    const discoveryTokenCached = client.getDiscoveryToken();
    if (discoveryTokenCached) {
      console.log('\nDiscovery Token:');
      console.log(`  Token: ${discoveryTokenCached.token.substring(0, 50)}...`);
      console.log(`  Expires at: ${discoveryTokenCached.expires_at}`);
    }

    if (domainId) {
      const domainAccessCached = client.getCachedDomainAccess(domainId);
      if (domainAccessCached) {
        console.log('\nDomain Access:');
        console.log(`  Domain: ${domainAccessCached.name} (${domainAccessCached.id})`);
        console.log(`  Server: ${domainAccessCached.domain_server.name}`);
        console.log(`  Server URL: ${domainAccessCached.domain_server.url}`);
        console.log(`  Access token: ${domainAccessCached.access_token.substring(0, 50)}...`);
      }
    }

    // Save state example
    console.log('\n=== State Serialization ===');
    const stateJson = client.saveState();
    console.log(`State serialized (${stateJson.length} bytes)`);
    console.log('Note: Credentials are NOT included in saved state');

    // Demonstrate restoring from state
    console.log('\n=== State Restoration ===');
    const restoredClient = Client.fromState(stateJson, config);
    console.log('✓ Client restored from saved state');
    console.log(`  Is authenticated: ${restoredClient.isAuthenticated()}`);

    // Clean up expired tokens
    restoredClient.validateState();
    console.log('✓ Validated state (removed expired tokens if any)');

    console.log('\n✓ Example completed successfully!');

  } catch (error) {
    console.error('\n❌ Authentication failed:', error.message);
    if (error.retryable) {
      console.error('  This error is retryable');
    }
    process.exit(1);
  }
}

// Run the example
main().catch((error) => {
  console.error('💥 Unexpected error:', error);
  process.exit(1);
});

File: modules/authentication/pkg/python/examples/basic.py
#!/usr/bin/env python3
"""
Basic Authentication Example

This example demonstrates the complete authentication flow:
- Network authentication
- Discovery service authentication
- Domain access
- State persistence
"""
import asyncio
import os
from dotenv import load_dotenv
from auki_authentication import (
    Client,
    Config,
    AuthenticationError,
)


async def main():
    print("=== Auki Authentication Example ===\n")

    # Load environment variables
    load_dotenv(dotenv_path="../../.env")

    # Configuration
    # Note: refresh_url is included in api_url/refresh endpoint
    config = Config(
        api_url=os.getenv("API_URL", "https://api.aukiverse.com"),
        refresh_url=os.getenv("REFRESH_URL", "https://api.aukiverse.com/user/refresh"),
        dds_url=os.getenv("DDS_URL", "https://dds.posemesh.org"),
        client_id=os.getenv("CLIENT_ID", "python-sdk"),
        refresh_threshold_ms=int(os.getenv("REFRESH_THRESHOLD_MS", "300000"))
    )

    print("Configuration:")
    print(f"  API URL: {config.api_url}")
    print(f"  Refresh URL: {config.refresh_url}")
    print(f"  DDS URL: {config.dds_url}")
    print(f"  Client ID: {config.client_id}")
    print(f"  Refresh threshold: {config.refresh_threshold_ms}ms\n")

    # Credentials
    credentials = {
        'type': 'email',
        'email': os.getenv("EMAIL"),
        'password': os.getenv("PASSWORD")
    }

    if not credentials['email'] or not credentials['password']:
        print("❌ EMAIL and PASSWORD must be set in .env file")
        return

    print("Using Email/Password authentication")
    print(f"  Email: {credentials['email']}\n")

    # Create the authentication client
    async with Client(config, credentials) as client:
        # Check initial authentication state
        is_auth = client.is_authenticated()
        print(f"Initial authentication state: {'Authenticated' if is_auth else 'Not authenticated'}")
        print()

        try:
            # Step 1: Authenticate to the Auki network
            print("=== Step 1: Network Authentication ===")
            network_token = await client.authenticate()
            print("✓ Network authentication successful!")
            print(f"  Token expires at: {network_token.expires_at}")
            print()

            # Step 2: Authenticate to Discovery service
            print("=== Step 2: Discovery Authentication ===")
            discovery_token = await client.authenticate_discovery()
            print("✓ Discovery authentication successful!")
            print(f"  Token expires at: {discovery_token.expires_at}")
            print()

            # Step 3: Get domain access
            domain_id = os.getenv("DOMAIN_ID")
            if not domain_id:
                print("⚠ DOMAIN_ID not set in .env, skipping domain access step")
            else:
                print("=== Step 3: Domain Access ===")
                print(f"Requesting access to domain: {domain_id}")

                domain_access = await client.get_domain_access(domain_id)
                print("✓ Domain access granted!")
                print(f"  Domain ID: {domain_access.id}")
                print(f"  Domain Name: {domain_access.name}")
                print(f"  Server URL: {domain_access.domain_server.url}")
                print(f"  Server Name: {domain_access.domain_server.name}")
                print(f"  Server Region: {domain_access.domain_server.cloud_region}")
                print(f"  Location: {domain_access.domain_server.latitude}, {domain_access.domain_server.longitude}")
                print(f"  Access token: {domain_access.access_token[:50]}...")
                print(f"  Token expires at: {domain_access.expires_at}")
                print()

            # Display final state
            print("=== Final State ===")
            print(f"Is authenticated: {client.is_authenticated()}")

            network_token_cached = client.get_network_token()
            if network_token_cached:
                print("\nNetwork Token:")
                print(f"  Token: {network_token_cached.token[:50]}...")
                print(f"  Expires at: {network_token_cached.expires_at}")

            discovery_token_cached = client.get_discovery_token()
            if discovery_token_cached:
                print("\nDiscovery Token:")
                print(f"  Token: {discovery_token_cached.token[:50]}...")
                print(f"  Expires at: {discovery_token_cached.expires_at}")

            if domain_id:
                domain_access_cached = client.get_cached_domain_access(domain_id)
                if domain_access_cached:
                    print("\nDomain Access:")
                    print(f"  Domain: {domain_access_cached.name} ({domain_access_cached.id})")
                    print(f"  Server: {domain_access_cached.domain_server.name}")
                    print(f"  Server URL: {domain_access_cached.domain_server.url}")
                    print(f"  Access token: {domain_access_cached.access_token[:50]}...")

            # Save state example
            print("\n=== State Serialization ===")
            state_json = client.save_state()
            print(f"State serialized ({len(state_json)} bytes)")
            print("Note: Credentials are NOT included in saved state")

            # Demonstrate restoring from state
            print("\n=== State Restoration ===")
            restored_client = Client.from_state(state_json, config)
            print("✓ Client restored from saved state")
            print(f"  Is authenticated: {restored_client.is_authenticated()}")

            # Clean up expired tokens
            restored_client.validate_state()
            print("✓ Validated state (removed expired tokens if any)")

            print("\n✓ Example completed successfully!")

        except AuthenticationError as error:
            print(f"\n❌ Authentication failed: {error}")
            if error.retryable:
                print("  This error is retryable")


if __name__ == "__main__":
    asyncio.run(main())
